certbot-apache/tests/augeasnode_test.py

```diff
@@ -1,15 +1,23 @@
 """Tests for AugeasParserNode classes"""
 import mock
 
+import unittest
 import util
 
+try:
+    import apacheconfig  # pylint: disable=import-error,unused-import
+    HAS_APACHECONFIG = True
+except ImportError:  # pragma: no cover
+    HAS_APACHECONFIG = False
+
 from acme.magic_typing import List  # pylint: disable=unused-import, no-name-in-module
 from certbot import errors
 
 from certbot_apache._internal import assertions
 
 
 
+@unittest.skipIf(not HAS_APACHECONFIG, reason='Tests require apacheconfig dependency')
```

Comment:
i still think we might need a `HAS_APACHECONFIG` flag for tests to indicate whether `apacheconfig` is present in the env, since these tests will still get run with and without apacheconfig in the env, and should work in both cases

Reply:
Wait a second, why is the `AugeasNode` test relying on `apacheconfig` at all? That's surprising.

Reply:
The AugeasNode test actually performs the tests on the DualNodes-- so as we start fleshing out the apacheconfig side, it will run that code as well. It's actually been quite useful since I haven't had to write tests for the apacheconfig side-- these tests coverage most all the functionality.

Further down the line, I rename these tests/the files-- we can do that here in this PR instead.

Reply:
Ah. So as it is now, the tests would work without `apacheconfig`? I'd rather only include the skip when it's in context with the change that requires it. I think I'd rather hold off and see it all in context together; maybe there's a cleaner change than env gating in the future. Is this the only reason to have the env gating up in `configurator.py` as well?

Reply:
DualNode calls the constructor for both the AugeasNode and the ApacheNode, we do create the parser tree in both contexts. We could have the env gating within DualNode instead-- and have it only create the ApacheNode if apacheconfig exists.

Reply:
Ok, I've been trying to figure out what's setting off my alarm bells. I think it's a combination of these things:
- Having another thing to remember to change when we push it live
- Adding inconsistent/special-cased tests
- Making it harder to figure out which code was running in which test

My proposal to get the same benefits of the gating without running into these problems:
- Add `certbot-apache[dev]` to `all_packages` in `tox.ini` so it's present for all tests
- When it's time to go public, move `apacheconfig` out of dev (which we'll have to do anyway)

Benefit:
- Since `USE_PARSERNODE` is only set in testing environments, we'll never run that code without `apacheconfig` being present.

Drawback?
- Did we plan to release the `AugeasNode` implementation without releasing the `ApacheNode` implementation? If we did, we'd have to do what you suggested and gate instantiating `ApacheNode` on `apacheconfig` being present, but then we shouldn't need to add the `skipif` to the broader `AugeasNode` tests, only the ones that specifically rely on `apacheconfig`. But also I don't think we planned to do this, though I could be mistaken.

Reply:
> Add `certbot-apache[dev]` to `all_packages` in `tox.ini` so it's present for all tests

Then the test suites wouldn't be run for `certbot-apache` when it doesn't have `apacheconfig` in the dev, which is what we're releasing to people. Is that okay?

> Did we plan to release the AugeasNode implementation without releasing the ApacheNode implementation

I don't think so. 

> If we did, we'd have to do what you suggested and gate instantiating ApacheNode on apacheconfig being present, but then we shouldn't need to add the skipif to the broader AugeasNode tests, only the ones that specifically rely on apacheconfig. But also I don't think we planned to do this, though I could be mistaken.

The AugeasNode tests are still a misnomer though-- they're only currently testing "Augeas nodes" since we haven't implemented "ApacheNode" yet-- once we've implemented ApacheNode, they'll all rely on `apacheconfig`.

Reply:
Continuing the conversation here for context. Seeing how it is now with these changes, I don't actually think `AugeasNode` was a misnomer at all. These tests are running through the functionality for augeas nodes, and happened to have a weird side effect of also being useful for testing `ApacheNode`s. Do we not want to have parallel tests for `ApacheNode`s instead of relying on these tests? Looking more closely at the breakdown of what's tested in the tests makes me think even more that the `AugeasNode` tests shouldn't depend on `apacheconfig`.

Reply:
I'm fine with that too. To make this work I mocked out DualNode into AugeasNode, and removed references to `primary`. ready for another go!